File: Lab-6/Algorithms C++/AlgorithmsV2/Algorithms.cpp
#include "Algorithms.h"
#include <limits.h>
#include <iostream>

using namespace std;


tuple<int, int> Algorithms::FindMinInArray(int* array, int size)
{
	int min = INT_MAX;
	int index = 0;

	for (int i = 0; i < size; i++)
	{
		if (array[i] < min)
		{
			min = array[i];
			index = i;
		}
	}

	return tuple<int, int>(min, index);
}

int* Algorithms::BubbleSort(int* array, int size)
{
	for (int i = 0; i < size; i++)
	{
		for (int j = i + 1; j < size; j++)
		{
			if (array[i] > array[j])
			{
				int temp = array[i];
				array[i] = array[j];
				array[j] = temp;
			}
		}
	}

	return array;
}


int Algorithms::BinarySearch(int array[], int size, int elem)
{
	int left = 0, right = size, mid = 0;

	while (left <= right)
	{
		mid = left + (right - left) / 2;
		if (elem < array[mid])
			right = mid - 1;
		else
			if (elem > array[mid])
				left = mid + 1;
			else 
				return mid;
	}

	return -1;
}


tuple<int, int> Algorithms::FindMinInMatrix(int(*matrix)[4], int rows)
{
	int columns = sizeof(*matrix) / sizeof(*matrix[0]);
	int min = INT_MAX;
	int index = 0;

	for (int i = 0; i < rows; i++)
	{
		for (int j = 0; j < columns; j++)
		{
			if (matrix[i][j] < min)
			{
				min = matrix[i][j];
				index = i;
			}
		}
	}

	return tuple<int, int>(min, index);
}


int* Algorithms::ReverseArray(int* array, int size)
{
	for (int i = 0, j = size - 1; i < j; i++, j--)
	{
		int t = array[i];
		array[i] = array[j];
		array[j] = t;
	}

	return array;
}


int* Algorithms::ShiftArrayElements(int* array, int size, int shiftParameter)
{
	int *newArray = new int[size];

	for (int i = 0; i < size; i++)
	{
		newArray[i] = array[(i + shiftParameter) % size];
	}

	return newArray;
}


int* Algorithms::ReplaceElements(int* array, int size, int elem, int value)
{
	for (int i = 0; i < size; i++)
	{
		if (array[i] == elem)
		{
			array[i] = value;
		}
	}

	return array;
}

File: Lab-6/Algorithms C#/Algorithms/Algorithms.cs
using System;


namespace Algorithms
{
    class Algorithms
    {
        public (int, int) FindMinInArray(int[] array)
        {
            int min = int.MaxValue;
            int index = 0;

            for (int i = 0; i < array.Length; i++)
            {
                if (array[i] < min)
                {
                    min = array[i];
                    index = i;
                }
            }

            return (min, index);
        }


        public int[] BubbleSort(int[] array)
        {
            for (int i = 0; i < array.Length; i++)
            {
                for (int j = i + 1; j < array.Length; j++)
                {
                    if (array[i] > array[j])
                    {
                        int temp = array[i];
                        array[i] = array[j];
                        array[j] = temp;
                    }
                }
            }

            return array;
        }


        public int BinarySearch(int[] array, int elem)
        {
            return Array.BinarySearch(array, elem);
        }


        public (int, int) FindMinInMatrix(int[,] matrix)
        {
            int rows = matrix.GetUpperBound(0) + 1;
            int columns = matrix.Length / rows;
            int min = int.MaxValue;
            int index = 0;

            for (int i = 0; i < rows; i++)
            {
                for (int j = 0; j < columns; j++)
                {
                    if (matrix[i, j] < min)
                    {
                        min = matrix[i, j];
                        index = i;
                    }
                }
            }

            return (min, index);
        }


        public int[] ReverseArray(int[] array)
        {
            for (int i = 0, j = array.Length - 1; i < j; i++, j--)
            {
                var t = array[i];
                array[i] = array[j];
                array[j] = t;
            }

            return array;
        }


        public int[] ShiftArrayElements(int[] array, int shiftParameter)
        {
            int[] newArray = new int[array.Length];

            for (int i = 0; i < array.Length; i++)
            {
                newArray[i] = array[(i + shiftParameter) % newArray.Length];
            }

            return newArray;
        }


        public int[] ReplaceElements(int[] array, int elem, int value)
        {
            for (int i = 0; i < array.Length; i++)
            {
                if (array[i] == elem)
                {
                    array[i] = value;
                }
            }

            return array;
        }
    }
}
